### PR TITLE
Expose the length and type of memory blocks.

### PR DESCRIPTION
**Pull request by ulysseB**
This PR exposes the length and element type of memory block in code generation instead of just providing a number of bytes to allocate.

Suggested review order:
- `src/ir/mem.rs`
- `src/codegen/function.rs`
- all the rest.

**Do not merge into master, merge in the variable branch instead**

### CODE REVIEW

File: src/ir/mem.rs [hunk 1]
```diff
@@ -59,9 +59,14 @@ impl Block {
         true
     }
 
-    /// Return the base size of the block, if it is statically known.
-    pub fn base_size(&self) -> u32 {
-        self.base_size * unwrap!(self.elements_type.len_byte())
+    /// Return the base size of the block in bytes.
```
Comment (nicoTolly) [inline src/ir/mem.rs]: What do you mean by "base size" ? Is it just the number of bytes in the block or is there any other subtlety ? If it is simply the number of bytes, remove the word base

File: src/codegen/function.rs [hunk 2]
```diff
@@ -331,25 +331,41 @@ impl<'a> MemoryRegion<'a> {
...
         if let Some(ref s) = self.num_private_copies {
             out *= s
         }
         out
     }
 
-    /// Returns the size of the part of the allocated memory accessible by each thread.
-    pub fn local_size(&self) -> &codegen::Size<'a> {
-        &self.size
+    /// Retruns the size of the memory block in number of elements accessible by each
```
Comment (nicoTolly) [inline src/codegen/function.rs]: Typo :)

File: src/ir/mem.rs (hunk 1 above)
Comment (ulysseB) [reply src/ir/mem.rs]: Added a comment to clarify.

**Comment by ulysseB** [comment]
bors retry

**Pull request merged**